source/common/network/socket_interface_singleton.h

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include "envoy/network/socket.h"
+
+#include "common/singleton/threadsafe_singleton.h"
+
+namespace Envoy {
+namespace Network {
+
+SocketInterface* createDefaultSocketInterface();
+
+template <class T> class CustomInjectableSingleton {
```

Comment:
I don't think you need this class, and a `std::unique_ptr` static member is likely to be a cause of test flakiness due to the static initialization fiasco.

It seems you want to inject another socket interface in your own build, is that the case? If so see what has been done for envoy.common.crypto.utility_lib in envoy and envoy-openssl, then you should be able to inject it at linking phase.

Reply:
Yes, the goal with this is/was to support the injecting of a different socket interface but at runtime and by means of some startup config. Plan was to add that interface somewhere in extensions and to conditionally build/link it against the actual library to avoid slowing down the build for everybody. Is that not recommended, i.e., should that code reside outside of envoy? 

Independent of all this, I assume we still want the singleton for all those situations when a component wants to directly interact with the socket layer (instead of using sockets and their apis) without a handle to the interface. Don't we?

Reply:
> Yes, the goal with this is/was to support the injecting of a different socket interface but at runtime and by means of some startup config. Plan was to add that interface somewhere in extensions and to conditionally build/link it against the actual library to avoid slowing down the build for everybody. Is that not recommended, i.e., should that code reside outside of envoy?

OK if you want to do this in config time, then probably I'd want you to wait for #11219 (which should be posted in PR today or tomorrow). 

> 
> Independent of all this, I assume we still want the singleton for all those situations when a component wants to directly interact with the socket layer (instead of using sockets and their apis) without a handle to the interface. Don't we?

I mean, you need Singleton, but you shouldn't need a `CustomInjectableSingleton` class, Combining existing `InjectableSingleton`/`ScopedInjectableLoader` should be able to what you want.

Reply:
> > Yes, the goal with this is/was to support the injecting of a different socket interface but at runtime and by means of some startup config. Plan was to add that interface somewhere in extensions and to conditionally build/link it against the actual library to avoid slowing down the build for everybody. Is that not recommended, i.e., should that code reside outside of envoy?
> 
> OK if you want to do this in config time, then probably I'd want you to wait for #11219 (which should be posted in PR today or tomorrow).

Not a problem! I don't yet have this code ready either way, as I'm still doing things in small steps. Mainly to help with the reviews and to avoid breaking things but also because I'm not yet familiarized with many parts of Envoy. 

> 
> > Independent of all this, I assume we still want the singleton for all those situations when a component wants to directly interact with the socket layer (instead of using sockets and their apis) without a handle to the interface. Don't we?
> 
> I mean, you need Singleton, but you shouldn't need a `CustomInjectableSingleton` class, Combining existing `InjectableSingleton`/`ScopedInjectableLoader` should be able to what you want.

Ow. I actually wanted to do that but failed because tests don't always initialize the Api, so they end up crashing (hence `createDefaultSocketInterface`). Is there a better place where to initialize the `ScopedInjectableLoader`? I chose `Api:Impl` because it seemed the natural place to me ...

Reply:
> Ow. I actually wanted to do that but failed because tests don't always initialize the Api, so they end up crashing (hence `createDefaultSocketInterface`). Is there a better place where to initialize the `ScopedInjectableLoader`? I chose `Api:Impl` because it seemed the natural place to me ...

For now I think you can initialize it in global static pointer like what crypto_lib does [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/common/crypto/utility_impl.cc#L101) in the default implementation. If this causes test failure let me know.

Reply:
Should've thought of that! Thanks, @lizan !

Would it be fine if, in the future, we override the global static pointer in Api:Impl through configuration?

Reply:
> Would it be fine if, in the future, we override the global static pointer in Api:Impl through configuration?

That should be doable through #11413 if you register an bootstrap extension.

Reply:
Great! Will give it a read. Thanks!